src/number_theory/lucas_primality.lean

```diff
@@ -0,0 +1,103 @@
+Authors: Bolton Bailey
+-/
+
+import data.zmod.basic
+import data.fintype.basic
+import field_theory.finite.basic
+import group_theory.order_of_element
+
+/-!
+# The Lucas test for primes.
+
+This file implements the Lucas test for primes (not to be confused with the Lucas-Lehmer test for
+Mersenne primes). A number a witnesses that n is prime if a has order n-1 in the multiplicative
+group of integers mod n. This is checked by verifying that a^(n-1) = 1 (mod n) and a^d ≠ 1 (mod n)
+for any divisor d | n. This test is the basis of the Pratt primality certificate.
+
+## TODO
+
+- Bonus: Show the reverse implication i.e. if a number is prime then it has a Lucas witness.
+  See [this stackexchange](https://math.stackexchange.com/a/59911/165144) answer for proof.
+- Write a tactic that uses this theorem to generate Pratt primality certificates
+- Integrate this into the norm_num primality verifier
+
+## Implementation notes
+
+It was a challenge to avoid diamonds when writing this file. Ultimately, I settled on introducing
+an instance of `fact (0 < p)` rather than making an instance of `fintype` on `units (zmod n)` as
+Mario showed how to do:
+
+```
+noncomputable instance units_zmod.fintype : Π n, fintype (units (zmod n))
+| 0     := units_int.fintype
+| (n+1) := units.fintype
+```
+
+Note that Mario also remarked this could be made computable.
+
+-/
+
+variables (p : ℕ) [fact (0 < p)]
+
+/--
+If a^r = 1 mod p, but a^(r/q) ≠ 1 mod p for all prime factors q of r, then a has order r in the
+multiplicative group mod p.
+-/
+theorem order_from_pows (n r : ℕ) (a : zmod n)
+  (hr : 0 < r) (ha : a^r = 1) (hd : (∀ q : ℕ, (nat.prime q) -> (q ∣ r) -> a^(r/q) ≠ 1)) :
+  order_of a = r :=
+begin
+  -- Let b be r/(order_of a), and show b = 1
+  cases exists_eq_mul_right_of_dvd (order_of_dvd_of_pow_eq_one ha) with b hb,
+  suffices : b = 1, by simp [this, hb],
+  -- Assume b is not one...
+  by_contra,
+  have b_min_fac_dvd_p_sub_one : b.min_fac ∣ r,
+    { have c_dvd_factor : ∃ (c : ℕ), b = c * b.min_fac,
+        exact exists_eq_mul_left_of_dvd b.min_fac_dvd,
+      cases c_dvd_factor with c hc,
+      rw [hc, ←mul_assoc] at hb,
+      exact dvd.intro_left (order_of a * c) (eq.symm hb), },
+  -- Use the minimum prime factor of b as q.
+  refine hd b.min_fac (nat.min_fac_prime h) b_min_fac_dvd_p_sub_one _,
+  rw [←order_of_dvd_iff_pow_eq_one, nat.dvd_div_iff (b_min_fac_dvd_p_sub_one),
+      hb, mul_comm, nat.mul_dvd_mul_iff_left (order_of_pos' _)],
+  { exact nat.min_fac_dvd b, },
+  { rw is_of_fin_order_iff_pow_eq_one,
+    exact Exists.intro r (id ⟨sub_pos_iff_lt.mpr hr, ha⟩), },
+end
+
+/--
+If a^(p-1) = 1 mod p, but a^((p-1)/q) ≠ 1 mod p for all prime factors q of p-1, then p is prime.
+This is true because a has order p-1 in the multiplicative group mod p, so this group must itself
+have order p-1, which only happens when p is prime.
+-/
+theorem lucas_primality (a : zmod p)
+  (hp : 1 < p) (ha : a^(p-1) = 1)
```

Comment:
This lemma is also true for `p = 1`, right? Because `hd` will be false (take `q = 2`).

Reply:
Well, that's one of these confusing things about zmod 1, it only works because 0 = 1. I did it this way because it seems like it would be confusing to people who come to the theorem later to reason about this.

Reply:
In my mind, it's more likely that someone would be trying to encounter problems thinking "Wait was it `0 < p` or `1 < p`, I guess its probably 1 < p because it doesn't make sense to compare something to 1 in zmod 1", than for someone to want to apply this theorem when p = 1, since the theorem is vacuous in that case anyway. 

If you feel strongly that the most general version of the theorem should be proved here then I can change it.

Reply:
I might be an autistic pedant, but for me it's more confusing with the superfluous hypothesis because I don't know why it's there and I can easily see that it doesn't need to be. Your "it only works because 0=1" is my "it works because obviously 0=1"

Reply:
It might be because I'm an autistic pedant, but my feelings are the exact opposite to yours: I don't see any reason why it "doesn't make sense to compare something to 1 in zmod 1" -- are you using the "schoolchild" model of zmod n as the set {0,1,...,n-1} (so `n : zmod n` "has no meaning") as opposed to the quotient model (which Lean is using, so `n : zmod n` has a perfectly good meaning)?